contracts/contracts/Multisig.cairo

```diff
@@ -470,24 +432,119 @@ func execute_transaction{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+# Sets number of required confirmations. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_confirmations_required.
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    require_multisig()
+
+    const lower_bound = 1
+
+    let (owners_len) = _owners_len.read()
+    with_attr error_message("invalid number of required confirmations"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
+    end
+
+    _set_confirmations_required_impl(confirmations_required)
+
+    return ()
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    require_multisig()
+    _set_owners_impl(owners_len, owners)
+
+    return ()
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
+    set_confirmations_required(confirmations_required)
+
+    return ()
+end
+
 #
 # Storage Helpers
 #
 
-func _set_owners{
-        syscall_ptr : felt*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr,
-    }(
-        owners_index : felt,
-        owners_len : felt,
-        owners : felt*,
-    ):
+func _get_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_index : felt, owners_len : felt, owners : felt*
+):
+    if owners_index == owners_len:
+        return ()
+    end
+
+    let (owner) = _owners.read(index=owners_index)
+    assert owners[owners_index] = owner
+
+    _get_owners(owners_index=owners_index + 1, owners_len=owners_len, owners=owners)
+    return ()
+end
+
+func _set_owners_impl{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    alloc_locals
+    require_unique_owners(owners_len, owners)
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
```

Comment:
Also our condition should Ideally be `owners_len < confirmations_required` instead of `owners_len <= confirmations_required`

Reply:
> Also our condition should Ideally be owners_len < confirmations_required instead of owners_len <= confirmations_required

That's exactly what is done here: is_le checks if `a <= b` so `owners_len <= confirmations_required - 1` with integers is exactly `owners_len < confirmations_required`. Unfortunately there's no `is_lt` in cairo lib.

Reply:
Ok, that makes sense. Consider resolved!

But also consider adding a short inline comment explaining this logic, as it will make it easier to read/reason.